File: wilbur-api/src/models/moderation.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type, PartialEq)]
#[sqlx(type_name = "report_status", rename_all = "lowercase")]
#[serde(rename_all = "lowercase")]
pub enum ReportStatus {
    Pending,
    Reviewed,
    Dismissed,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct BannedUser {
    pub id: Uuid,
    pub room_id: Uuid,
    pub user_id: Uuid,
    pub banned_by: Uuid,
    pub reason: Option<String>,
    pub expires_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct ModerationLog {
    pub id: Uuid,
    pub room_id: Uuid,
    pub moderator_id: Uuid,
    pub target_user_id: Uuid,
    pub action: String,
    pub details: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct ReportedContent {
    pub id: Uuid,
    pub room_id: Uuid,
    pub reporter_id: Uuid,
    pub content_type: String,
    pub content_id: Uuid,
    pub reason: String,
    pub status: ReportStatus,
    pub reviewed_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct BanUserRequest {
    pub user_id: Uuid,
    #[validate(length(max = 1000))]
    pub reason: Option<String>,
    pub expires_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Deserialize, Validate)]
pub struct ReportContentRequest {
    pub content_type: String,
    pub content_id: Uuid,
    #[validate(length(min = 1, max = 1000))]
    pub reason: String,
}

#[derive(Debug, Deserialize)]
pub struct ReviewReportRequest {
    pub status: ReportStatus,
}

/// Banned user response.
#[derive(Debug, Serialize)]
pub struct BannedUserResponse {
    pub id: Uuid,
    pub room_id: Uuid,
    pub user_id: Uuid,
    pub banned_by: Uuid,
    pub reason: Option<String>,
    pub expires_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
}

impl From<BannedUser> for BannedUserResponse {
    fn from(b: BannedUser) -> Self {
        Self {
            id: b.id,
            room_id: b.room_id,
            user_id: b.user_id,
            banned_by: b.banned_by,
            reason: b.reason,
            expires_at: b.expires_at,
            created_at: b.created_at,
        }
    }
}

/// Moderation log response.
#[derive(Debug, Serialize)]
pub struct ModerationLogResponse {
    pub id: Uuid,
    pub room_id: Uuid,
    pub moderator_id: Uuid,
    pub target_user_id: Uuid,
    pub action: String,
    pub details: Option<String>,
    pub created_at: DateTime<Utc>,
}

impl From<ModerationLog> for ModerationLogResponse {
    fn from(l: ModerationLog) -> Self {
        Self {
            id: l.id,
            room_id: l.room_id,
            moderator_id: l.moderator_id,
            target_user_id: l.target_user_id,
            action: l.action,
            details: l.details,
            created_at: l.created_at,
        }
    }
}

/// Reported content response.
#[derive(Debug, Serialize)]
pub struct ReportedContentResponse {
    pub id: Uuid,
    pub room_id: Uuid,
    pub reporter_id: Uuid,
    pub content_type: String,
    pub content_id: Uuid,
    pub reason: String,
    pub status: ReportStatus,
    pub reviewed_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

impl From<ReportedContent> for ReportedContentResponse {
    fn from(r: ReportedContent) -> Self {
        Self {
            id: r.id,
            room_id: r.room_id,
            reporter_id: r.reporter_id,
            content_type: r.content_type,
            content_id: r.content_id,
            reason: r.reason,
            status: r.status,
            reviewed_by: r.reviewed_by,
            created_at: r.created_at,
        }
    }
}

File: react-app/src/api/moderation.ts
/**
 * Moderation API module.
 */

import { api } from './client';

interface BannedUser {
  id: string;
  user_id: string;
  room_id: string;
  banned_by: string;
  reason: string | undefined;
  expires_at: string | undefined;
  created_at: string;
}

interface ModerationLogEntry {
  id: string;
  room_id: string;
  moderator_id: string;
  target_user_id: string;
  action: string;
  details: string | undefined;
  created_at: string;
}

interface Report {
  id: string;
  room_id: string;
  reporter_id: string;
  reported_user_id: string;
  content_type: string | undefined;
  content_id: string | undefined;
  reason: string;
  status: string;
  reviewed_by: string | undefined;
  created_at: string;
}

export const moderationApi = {
  ban(userId: string, roomId: string, reason?: string, durationSecs?: number): Promise<BannedUser> {
    return api.post<BannedUser>('/api/v1/moderation/ban', {
      user_id: userId,
      room_id: roomId,
      reason,
      duration_secs: durationSecs,
    });
  },

  unban(userId: string, roomId: string): Promise<void> {
    return api.post('/api/v1/moderation/unban', { user_id: userId, room_id: roomId });
  },

  kick(userId: string, roomId: string, reason?: string): Promise<void> {
    return api.post('/api/v1/moderation/kick', { user_id: userId, room_id: roomId, reason });
  },

  mute(userId: string, roomId: string, durationSecs?: number): Promise<void> {
    return api.post('/api/v1/moderation/mute', {
      user_id: userId,
      room_id: roomId,
      duration_secs: durationSecs,
    });
  },

  getLog(roomId: string): Promise<{ room_id: string; entries: ModerationLogEntry[] }> {
    return api.get(`/api/v1/moderation/log/${roomId}`);
  },

  getBannedUsers(roomId: string): Promise<{ room_id: string; banned_users: BannedUser[] }> {
    return api.get(`/api/v1/moderation/banned/${roomId}`);
  },

  report(roomId: string, reportedUserId: string, reason: string, messageId?: string): Promise<Report> {
    return api.post<Report>('/api/v1/moderation/report', {
      room_id: roomId,
      reported_user_id: reportedUserId,
      reason,
      message_id: messageId,
    });
  },

  resolveReport(reportId: string, status?: string): Promise<Report> {
    return api.post<Report>(`/api/v1/moderation/report/${reportId}/resolve`, { status });
  },
};
